src/app/icd/client/CheckInHandler.cpp

```diff
@@ -109,8 +109,20 @@ CHIP_ERROR CheckInHandler::OnMessageReceived(Messaging::ExchangeContext * ec, co
 
     if (refreshKey)
     {
-        // TODO: A new CASE session should be established to re-register the client using a new key. The registration will happen in
-        // CASE session callback
+        RefreshKeySender * refreshKeySender = mpCheckInDelegate->OnKeyRefreshNeeded(clientInfo, mpICDClientStorage);
+        if (refreshKeySender == nullptr)
+        {
+            ChipLogError(ICD, "Key Refresh failed for node ID:" ChipLogFormatScopedNodeId,
+                         ChipLogValueScopedNodeId(clientInfo.peer_node));
+            return CHIP_NO_ERROR;
```

Comment:
is it intentional not to call mpCheckInDelegate->OnKeyRefreshDone(..);? since OnKeyRefreshNeeded has handled the release in error cases?

Reply:
Yes. I don't call it since the refreshKeySender failed to be allocated and hence need not be released. However, if the deleagte is to take corrective action from OnKeyRefreshDone, I think it is better to call OnKeyRefreshDone from here.